harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -116,6 +171,127 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
         # torch.backends.cudnn.deterministic = True
         # torch.backends.cudnn.benchmark = False
 
+    def report_training_metrics(self, steps_completed):
+        metrics = self._prepare_metrics(self.training_metrics)
+        metrics = self.context.distributed.broadcast(metrics)
+
+        for callback in self.callbacks.values():
+            callback.on_training_workload_end(
+                avg_metrics=metrics["avg_metrics"],
+                batch_metrics=metrics["batch_metrics"],
+            )
+
+        # Only report on the chief worker
+        if self.is_chief:
+            avg_metrics = metrics.get("avg_metrics", {})
+            batch_metrics = metrics.get("batch_metrics", [])
+
+            self.metric_writer.on_train_step_end(
+                self.steps_completed,
+                avg_metrics,
+                batch_metrics,
+            )
+            self.core_context.train.report_training_metrics(
+                steps_completed=steps_completed, metrics=avg_metrics, batch_metrics=batch_metrics
+            )
+
+        self.context.reset_reducers()
+        self.training_metrics = []
+
+    def report_validation_metrics(self, steps_completed, metrics):
+        if self.is_chief:
+            # Check if op complete
+            searcher_metric = self._validate_searcher_metric(metrics)
+            if self.searcher_op and self._steps_remaining() < 1:
+                self.searcher_op.report_completed(searcher_metric)
```

Comment:
question: do we want to enforce that all searcher-mandated validations also create checkpoints?